# Lab 2: From Raw Data to ML-Ready Data
## Data Cleaning & Exploratory Data Analysis — Telco Customer Churn

**Name:** Ritika

**Roll Number:** 023-24-0262

**GitHub Repository:** https://github.com/Ritika-Lund/Machine-learning-Lab.git

**Duration:** 3 Hours (independent, hands-on)

> This notebook is a **template**, not a tutorial. You already know Python, NumPy, Pandas, and Matplotlib from previous labs. Each section below states the objective and the questions you must answer — **you decide** which functions/plots are appropriate. Do not just run code without interpreting it: every visualization needs a written observation, and every cleaning decision needs a written justification.
>
> Fill in the empty code cells and the *Answer:* / *Observation:* placeholders directly in this notebook. Do not delete the Markdown headings — they are used for grading.

---

## Setup

Import the libraries you'll need and load the dataset. Use the Telco Customer Churn CSV provided for this lab.

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load the dataset
df = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")  


---

## Part 1 — The Business Problem

A telecommunications company is losing customers to competitors — this is called **customer churn**. The company wants to understand which customers are likely to leave, so it can intervene early. Before any model can be built, the raw data must be understood, audited for quality issues, and explored.

**Tasks**
1. In 2–3 sentences, state the business problem in your own words, and propose what the target variable likely is.
2. List 5 columns you expect to be useful predictors of churn, and 2 columns you expect to be useless as predictors — justify each choice in one line.
3. Write down 2 questions you personally want this data to answer about customers who churn.

**Answer 1:**
The telecommunications company wants to identify customers who are likely to leave so it can take action before they churn. The goal is to understand which customer and service characteristics may be associated with churn. The target variable is Churn, which indicates whether a customer has left the company (Yes or No).

**Answer 2:**
These are the features i think will be useful and useless,confirmation can be only given after performing EDA



Useful features 
1)Contract 
2)Tenure
3)Monthly charges
4)PaymentMethod
5)InternetService

Useless Features
1)customerID
2)Gender

**Answer 3:**
1)Are customers with month-to-month contracts more likely to churn(leave) than customers with one-year or two-year contracts?
2)How much customer tenure is connected to customers who churn or customers who stay?

---

## Part 2 — Exploring the Dataset Structure

Before cleaning or analyzing anything, get an accurate picture of the dataset's shape and contents. Running a command is not the same as understanding its output — you must interpret what you see.

*Concepts/functions you may find useful:* `df.shape`, `df.columns`, `df.dtypes`, `df.info()`, `df.head()`, `df.sample(n)`, `df[col].unique()`, `df[col].nunique()`, `df.describe(include='all')`

**Tasks**
4. How many customers and how many features are in this dataset? List the numerical features and the categorical features separately.
5. Identify any column(s) that are simply identifiers rather than predictive features, and any column(s) whose data type looks wrong for what it represents.
6. Pick 3 categorical columns and report their unique values — do any contain unexpected or inconsistent categories?

In [3]:
# Task 4 — dataset shape, column names, data types
print("No of rows and columns :")
print(df.shape)
# there are total 21 columns from these 19 are features 1 is identifier 1 is target variable.

print("\nColumn data types:")
print(df.dtypes)
#  numerical 
    # tenure
    # MonthlyCharges
# categorical
    # gender
    # Partner
    # Dependents
    # PhoneService
    # MultipleLines
    # InternetService
    # OnlineSecurity
    # OnlineBackup
    # DeviceProtection
    # TechSupport
    # StreamingTV
    # StreamingMovies
    # Contract
    # PaperlessBilling
    # PaymentMethod
# seniorCitizen is in binary category


No of rows and columns :
(7043, 21)

Column data types:
customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


In [4]:
# Task 5 — identify identifier columns / wrong data types
# customerId is identifier,it is not a meaningful predictive feature
df["TotalCharges"].dtype
# TotalCharges should be float but instead it is string which is wrong data type

<StringDtype(na_value=nan)>

In [5]:
# Task 6 — unique values in 3 categorical columns
print(df['Contract'].unique())
print(df['InternetService'].unique())
print(df['PaymentMethod'].unique())

<ArrowStringArray>
['Month-to-month', 'One year', 'Two year']
Length: 3, dtype: str
<ArrowStringArray>
['DSL', 'Fiber optic', 'No']
Length: 3, dtype: str
<ArrowStringArray>
[         'Electronic check',              'Mailed check',
 'Bank transfer (automatic)',   'Credit card (automatic)']
Length: 4, dtype: str


**Observations (Part 2):**
The dataset contains 7,043 customers and 21 columns. customerID is an identifier and Churn is the target variable. tenure and MonthlyCharges are numerical features, while most of the remaining features are categorical. SeniorCitizen is stored as an integer but represents a binary category. TotalCharges is numerical in meaning but is stored as object, making it a suspicious column that requires further investigation. The three categorical columns examined using unique() did not show obvious inconsistent categories.

---

## Part 3 — Investigating Data Quality

A dataset can look clean at a glance and still hide serious problems. You must actively search for missing values, duplicates, and suspicious values rather than assume there are none.

*Concepts/functions you may find useful:* `df.isnull().sum()`, `df.isnull().mean()*100`, `df.duplicated().sum()`, `df['id_col'].duplicated().sum()`, `df[col].value_counts()`

**Tasks**
7. Report missing values as both a count and a percentage per column. Which missing-value problems do you think require action, and which don't? Justify your reasoning.
8. Check for duplicate rows and duplicate customer IDs separately — are they the same issue or different issues?
9. Find at least one column with a suspicious, invalid, or inconsistently formatted value, and describe how you found it.

In [14]:
# Task 7 — missing values: count and percentage per column
missing_values_count=df.isnull().sum()
print(missing_values_count)
print("\npercentage of missing values:")
missing_values_percentage=(df.isnull().mean() * 100)
print(missing_values_percentage)
df["TotalCharges"].str.strip().eq('').sum()
blank_totalcharges = df['TotalCharges'].str.strip().eq('').sum()

blank_percentage = (blank_totalcharges / len(df)) * 100

print("Blank TotalCharges:", blank_totalcharges)
print("Percentage:", blank_percentage)

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

percentage of missing values:
customerID          0.0
gender              0.0
SeniorCitizen       0.0
Partner             0.0
Dependents          0.0
tenure              0.0
PhoneService        0.0
MultipleLines       0.0
InternetService     0.0
OnlineSecurity      0.0
OnlineBackup        0.0
DeviceProtection    0.0
TechSupport         0.0
StreamingTV         0.0
StreamingMovies     0.0
Contract            0.0
PaperlessBilling    0.0
PaymentMethod       0.0
MonthlyCharges      0.0
TotalCharges        0.0
Churn         

In [18]:
# Task 8 — duplicate rows vs duplicate IDs
duplicate_rows=df.duplicated().sum()
print("No of duplicate rows in this dataset :",duplicate_rows)
duplicate_customerIds=df['customerID'].duplicated().sum()
print("No of duplicate customerIDs in this dataset :",duplicate_customerIds)


No of duplicate rows in this dataset : 0
No of duplicate customerIDs in this dataset : 0


In [21]:
# Task 9 — investigate a suspicious / invalid value
print(df['TotalCharges'].tail())
df[df['TotalCharges'].str.strip() == '']

7038    1990.5
7039    7362.9
7040    346.45
7041     306.6
7042    6844.5
Name: TotalCharges, dtype: str


,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


**Answer 7 (which missing-value problems need action, and why):**
 Task 7 related observation
 The standard isnull() check reports no missing values in the dataset. However, further investigation of TotalCharges revealed 11 blank values
 These blanks are effectively missing information even though Pandas does not identify them as NaN
 This issue requires action because TotalCharges represents a numerical quantity and the blank values will interfere with converting the column to a numeric data type.

**Answer 8 (duplicate rows vs duplicate IDs — same issue or different?):**
this dataset contains 0 duplicated rows and 0 duplicate customerIDs.both check represent different issues because duplicate rows mean the entire record has been repeated ,while dupliacte customer id means same customer is appearning in multiple records.

**Answer 9 (suspicious value found and how):**
TotalCharges was identified as a suspicious column because it represents a numerical amount but has an object data type. Further inspection showed 11 records containing blank strings. These values were detected by stripping whitespace and checking for empty strings. The issue needs to be addressed before the column can be used as a numerical feature.

---

## Part 4 — Data Cleaning Decisions

Cleaning is a series of justified decisions, not a fixed recipe. For every issue you fix, you must be able to explain why you chose that method over an alternative.

*Concepts/functions you may find useful:* `pd.to_numeric(s, errors='coerce')`, `df.fillna(value)`, `df.dropna()`, `df.drop_duplicates()`, `df[col].str.strip()`, `df.astype(dtype)`

**Tasks**
10. For every data-quality issue you found in Part 3, record: Problem → Decision → Method Used → Reason.
11. Apply your cleaning decisions to produce an intermediate cleaned dataframe, and confirm (with code output) that each issue is actually resolved.

**Task 10 — Cleaning Decision Log**

| Problem | Decision | Method Used | Reason |
|---|---|---|---|
| _e.g. TotalCharges stored as text_ | _..._ | _..._ | _..._ |
| _..._ | _..._ | _..._ | _..._ |
| _..._ | _..._ | _..._ | _..._ |

In [ ]:
# Task 11 — apply your cleaning decisions here


In [ ]:
# Task 11 (continued) — confirm each issue is resolved (e.g. re-check isnull().sum(), duplicated().sum(), dtypes)


---

## Part 5 — Univariate EDA

Look at individual variables in isolation before looking at relationships. Every chart must be followed by a written interpretation — a graph with no explanation earns no credit.

*Concepts/functions you may find useful:* `df[col].plot(kind='hist')` / `plt.hist()`, `df.boxplot(column=col)`, `df[col].value_counts().plot(kind='bar')`

**Tasks**
12. Visualize the distribution of at least 2 numerical features and 2 categorical features. For each chart, write 1–2 sentences on what you learned.
13. Do any numerical features show outliers or a skewed distribution? What might explain this?

In [ ]:
# Task 12 — numerical feature 1: distribution plot


**Observation:** _(what did you learn from this chart?)_

In [ ]:
# Task 12 — numerical feature 2: distribution plot


**Observation:** _(what did you learn from this chart?)_

In [ ]:
# Task 12 — categorical feature 1: distribution plot


**Observation:** _(what did you learn from this chart?)_

In [ ]:
# Task 12 — categorical feature 2: distribution plot


**Observation:** _(what did you learn from this chart?)_

**Answer 13 (outliers / skew and possible explanation):**
_(your answer here)_

---

## Part 6 — Target Variable Analysis

The `Churn` column is what you will eventually predict. Before Lab 3, you need to understand its distribution — this has direct consequences for how a model should later be trained and evaluated.

*Concepts/functions you may find useful:* `df['Churn'].value_counts()`, `df['Churn'].value_counts(normalize=True)`

**Tasks**
14. Report the number and percentage of churned vs. non-churned customers. Is the target balanced or imbalanced?
15. In 2–3 sentences, explain why class imbalance might matter when a classification model is trained later — do not describe how to fix it, just why it matters.

In [ ]:
# Task 14 — churn counts and percentages


**Answer 14 (balanced or imbalanced?):**
_(your answer here)_

**Answer 15 (why imbalance might matter later):**
_(your answer here)_

---

## Part 7 — Bivariate Analysis: Features vs. Churn

Now connect individual features to the target. You choose the appropriate visualization or summary for each feature type — a categorical feature is not explored the same way as a numerical one.

*Concepts/functions you may find useful:* `pd.crosstab(df[col], df['Churn'], normalize='index')`, `df.groupby(col)['Churn']...`, `sns.boxplot(x='Churn', y=col, data=df)`

**Tasks**
16. Investigate at least 2 categorical features (e.g. Contract, InternetService, PaymentMethod) against Churn. For each, record: Observation / Evidence / Possible Explanation.
17. Investigate at least 2 numerical features (e.g. tenure, MonthlyCharges) against Churn using the same Observation / Evidence / Possible Explanation format.

In [ ]:
# Task 16 — categorical feature 1 vs Churn


**Observation:** _..._
**Evidence:** _..._
**Possible Explanation:** _..._

In [ ]:
# Task 16 — categorical feature 2 vs Churn


**Observation:** _..._
**Evidence:** _..._
**Possible Explanation:** _..._

In [ ]:
# Task 17 — numerical feature 1 vs Churn


**Observation:** _..._
**Evidence:** _..._
**Possible Explanation:** _..._

In [ ]:
# Task 17 — numerical feature 2 vs Churn


**Observation:** _..._
**Evidence:** _..._
**Possible Explanation:** _..._

---

## Part 8 — Correlation & Feature Relationships

Numerical features can be related to each other, not just to the target. Strongly correlated features can cause redundancy problems later, so it's worth knowing about them now.

*Concepts/functions you may find useful:* `df.corr(numeric_only=True)`, `sns.heatmap(corr_matrix, annot=True)`

**Tasks**
18. Compute and visualize the correlation matrix of the numerical features. Which pairs appear most correlated?
19. Does correlation imply causation here? Could any highly correlated features cause problems for a future model? Answer in 2–3 sentences.

In [ ]:
# Task 18 — correlation matrix and heatmap


**Answer 18 (most correlated pairs):**
_(your answer here)_

**Answer 19 (correlation vs. causation; potential problems):**
_(your answer here)_

---

## Part 9 — Data Leakage Awareness

In Lab 3 you will train a classifier. Before that, you must check whether any column contains information that would not realistically be available at the moment of prediction — using such a column would make a model look artificially accurate.

**Tasks**
20. Inspect the columns for anything that looks like it could only be known after a customer has already churned, or that is purely an identifier. List any suspects and explain your reasoning.
21. In 2–3 sentences, explain what would go wrong if such a column were left in the model's input features.

**Answer 20 (suspect columns and reasoning):**
_(your answer here)_

**Answer 21 (what would go wrong):**
_(your answer here)_

---

## Part 10 — ML-Readiness Checklist

Summarize every column's fate in a single table. This is the bridge between EDA and modeling — Lab 3 will use this table directly.

**Task**
22. Complete the ML-readiness table for **every** column in the dataset.

| Column | Type | Role | Action | Reason |
|---|---|---|---|---|
| customerID | Identifier | — | Remove | ... |
| tenure | Numerical | Feature | ... | ... |
| Contract | Categorical | Feature | ... | ... |
| Churn | Categorical | Target | Keep | ... |
| ... | ... | ... | ... | ... |

_(Add one row per remaining column — copy the pattern above.)_

---

## Part 11 — Building the Clean Dataset

Bring together every cleaning decision into one final dataframe, `clean_df`, that Lab 3 can load directly.

> Do **not** split into train/test, encode categorical variables, or train any model — that happens in Lab 3.

**Tasks**
23. Produce `clean_df` with correct data types, missing values handled, duplicates handled, and irrelevant identifier columns removed where justified. Save it (e.g. `clean_churn.csv`) alongside your notebook.
24. Print `df.info()` (or equivalent) for `clean_df` as final proof that every issue identified earlier has been resolved.

In [ ]:
# Task 23 — assemble clean_df from your cleaning decisions
clean_df = df.copy()

# ... apply your cleaning steps here ...

# Save the cleaned dataset
clean_df.to_csv("clean_churn.csv", index=False)


In [ ]:
# Task 24 — final proof: info() / isnull().sum() / duplicated().sum() on clean_df


---

## Part 12 — Key Findings & Final Reflection

Step back from the code and summarize what you actually learned about this dataset and this problem.

**Task 25 — Five Important Findings**

Repeat this block for each of your five findings.

**Finding 1**
- Observation: _..._
- Evidence: _..._
- Interpretation: _..._

**Finding 2**
- Observation: _..._
- Evidence: _..._
- Interpretation: _..._

**Finding 3**
- Observation: _..._
- Evidence: _..._
- Interpretation: _..._

**Finding 4**
- Observation: _..._
- Evidence: _..._
- Interpretation: _..._

**Finding 5**
- Observation: _..._
- Evidence: _..._
- Interpretation: _..._

---

**Task 26 — Final Reflection Questions**

1. What was the most important data-quality issue you found?
   _(your answer here)_
2. What was the most interesting pattern you discovered?
   _(your answer here)_
3. Which feature seems most useful for predicting churn?
   _(your answer here)_
4. Which feature should probably not be used, and why?
   _(your answer here)_
5. What preprocessing will be required before ML modeling (conceptually — do not implement it)?
   _(your answer here)_
6. What would you do next if you had another 3 hours?
   _(your answer here)_

---

## Bonus Tasks (Optional)

- Create one additional meaningful visualization not required above.
- Investigate a surprising customer subgroup you didn't expect to find.
- Calculate churn rate across a combination of 2+ customer segments at once.
- Identify one possible feature-engineering opportunity for Lab 3.
- Identify one possible source of bias in this dataset.

In [ ]:
# Bonus — optional exploration


---

## Submission Checklist

Before submitting, confirm you have:

- [ ] Completed all 26 tasks above, with written interpretation for every chart and decision
- [ ] A working `clean_df`, saved as `clean_churn.csv`
- [ ] A completed ML-readiness table (Part 10) covering every column
- [ ] Five Key Findings written (Part 12)
- [ ] Answered all six Final Reflection questions
- [ ] Committed and pushed this notebook + `clean_churn.csv` to your GitHub repository
- [ ] Added your GitHub repository link at the top of this notebook